### PR TITLE
Initial implementation of ContinuationLocal

### PR DESCRIPTION
**Pull request by eirslett**
> **Initial implementation of ContinuationLocal**
> The implementation is a direct port of com.twitter.util.Local:
> 
> https://github.com/twitter/util/blob/fe29be6ce62d27611ddfe94fdc9e9ec294375330/util-core/src/main/scala/com/twitter/util/Local.scala
> https://github.com/twitter/util/blob/fe29be6ce62d27611ddfe94fdc9e9ec294375330/util-core/src/test/scala/com/twitter/util/LocalTest.scala
> 
> An ExecutorService wrapper is also provided, meant to wrap thread
> pools so that the ContinuationLocal state is passed on. This
> implementation is inspired by BraveExecutorService and FuturePool:
> 
> https://github.com/kristofa/brave/blob/df55efd5653bccdb4736fd63413f10772a5bb5e2/brave-core/src/main/java/com/github/kristofa/brave/BraveExecutorService.java
> https://github.com/twitter/util/blob/fe29be6ce62d27611ddfe94fdc9e9ec294375330/util-core/src/main/scala/com/twitter/util/FuturePool.scala

Here's a (more or less) direct port of com.twitter.util.Local to Java, with only a few differences.
This ContinuationLocal needs mainstream adoption in order to work (every library that's dealing with threads and/or events needs to support it), so we have to get it right.

Any comments from the Twitter guys who wrote this (in Scala) in the first place?
@adriancole @mosesn @mariusae @evnm ?

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
@daniel-bryant-uk looking at your [not so recent blog](https://taidevcouk.wordpress.com/2014/05/26/implementing-correlation-ids-in-spring-boot/), you might be interested in this.

File: README.md [hunk 1]
```diff
@@ -1,2 +1,46 @@
...
+without any external dependencies (notably; no dependency on Scala).
+
+The ContinuationLocal can be used for any data that should be persisted in
+a continuation scope, but is especially useful for various monitoring and tracing
+implementations.
+
+## Usage example
+
+```java
+import com.github.distributivetracing.ContinuationLocal
```
Comment (codefromthecrypt) [inline README.md]: typo

File: README.md [hunk 2]
```diff
@@ -1,2 +1,46 @@
 # continuation-local-storage-jvm
-A common JVM implementation of continuation-local storage
+
+This is a common implementation of continuation-local storage for the JVM.
+It is based on Twitter's com.twitter.util.Local, but this is a Java port
```
Comment (codefromthecrypt) [inline README.md]: supernit, backtick type names? `com.twitter.util.Local` or even better link to it.

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        <groupId>org.sonatype.oss</groupId>
+        <artifactId>oss-parent</artifactId>
+        <version>7</version>
+    </parent>
+
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    </properties>
+
+    <groupId>com.github.distributedtracing</groupId>
```
Comment (codefromthecrypt) [inline pom.xml:15]: com.github probably isn't the right group id, as we won't be able to publish against that. @rtyler correct me if I am wrong on this.

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.12</version>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <pluginManagement>
+            <plugins>
```
Comment (codefromthecrypt) [inline pom.xml:79]: add animalsniffer?

``` xml
      <plugin>
        <groupId>org.codehaus.mojo</groupId>
        <artifactId>animal-sniffer-maven-plugin</artifactId>
        <version>1.14</version>
        <executions>
          <execution>
            <phase>test</phase>
            <goals>
              <goal>check</goal>
            </goals>
          </execution>
        </executions>
        <configuration>
          <signature>
            <groupId>org.codehaus.mojo.signature</groupId>
            <artifactId>java16</artifactId>
            <version>1.1</version>
          </signature>
        </configuration>
      </plugin>
```

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
+package com.github.distributedtracing;
+
+public final class ContinuationLocal<T> {
```
Comment (codefromthecrypt) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:3]: some things in the readme could be moved here I think.

File: pom.xml (hunk 3 above)
Comment (eirslett) [inline pom.xml:15]: There's a lot of packages out there released from github: http://search.maven.org/#search%7Cga%7C1%7Ccom.github

But maybe we should aquire a domain like "distributedtracing.org" and use "org.distributedtracing" as a groupId? Then we could point that domain to a gh-pages here in the github repo. (And the specification won't be tied to github)

**Comment by codefromthecrypt** [comment]
TL;DR;
I think a port of `com.twitter.util.Local` might be better in the openzipkin org, as that zipkin is the only JVM implementation that would be biased towards twitter commons idioms.

---

Disclaimer: this is just an opinion of a passer-by.. don't anchor too much on it. I'm just voicing my 2p.

---

I'm glad to see a proposal to a part of the [Shared JVM Implementation](https://docs.google.com/document/d/10S3wGxV-s2wFeXFE5ZrPAxYUW_HVJ-kUvxW6xG6_A74/edit) problem. My opinion, but I'd define success in this org as folks actually sharing a context vs being like `com.twitter.util.Local`, which only zipkin is interested in. Don't get me wrong.. I do think this is valuable, just guessing it is more valuable in the openzipkin org, as zipkin is the only [impl](https://drive.google.com/folderview?id=0B0tSnQT3uGdAflJLdEFhVzl6dEtrN0tPOFhmclFpOFJ5a01nZnFZaXdxdUJ2TUJfOGxhWUE&usp=drive_web) we can assume is familiar with `com.twitter.util.Local`.

Those who might use a java context propagation api, in terms of dapper, include HTrace (@saintstack), Brave (@kristofa), Pinpoint (@Xylus), Lyft (@jpinner), Spring Cloud (@dsyer @spencergibb) and Netflix (@NiteshKant). There's also work brewing in [GRPC](https://github.com/grpc/grpc-java/pull/562/files) which will eventually support dapper itself.

File: pom.xml (hunk 3 above)
Comment (codefromthecrypt) [inline pom.xml:15]: yeah I suppose I could have googled that :) I don't know what's best, but I suspect that not having github in package names or group ids could lead to later flexibility (thinking apache etc)

File: README.md (hunk 1 above)
Comment (eirslett) [inline README.md]: Oops. Referring to the missing semicolon?

Comment (codefromthecrypt) [inline README.md]: Distribu'v'e A couple places in the file.

**Comment by eirslett** [comment]
Yes, it's a port of `com.twitter.util.Local`, but that's just to have a starting point that is proven to work, not to be biased towards twitter commons in the long term (or even in the short term).
The reason why I think it belongs to distributedtracing in general, rather than openzipkin specifically, is that I think it will gain wider adoption from library authors (netflix oss, rxjava, finagle, akka...) if we keep the construct as agnostic as possible.
For example reactive streams, which is a very generic concept, would benefit from supporting continuation-local storage, but probably wouldn't support a Zipkin-specific construct because that's too implementation-specific.
I suppose Pinpoint could benefit from using this construct as well?

**Comment by eirslett** [comment]
> as zipkin is the only impl we can assume is familiar with com.twitter.util.Local.

It looks like most other implementations are familiar with `ThreadLocal` from Java, it's just that we have to sell `ContinuationLocal` as simply an extension to `ThreadLocal` that also supports thread pools and event loops?

File: pom.xml (hunk 3 above)
Comment (eirslett) [inline pom.xml:15]: Then I suggest that we buy that domain, and change the groupId.

File: README.md (hunk 1 above)
Comment (eirslett) [inline README.md]: ok, fixed!

File: pom.xml (hunk 3 above)
Comment (rtyler) [inline pom.xml:15]: com.github.distributedtracing is certainly legitimate and can be used for publishing. So can "distributedtracing" or anything in between.

Group IDs are only _conventionally_ supposed to be reverse DNS to the authors domain, and distributedtracing.github.com **is** a resolvable name and will point to the github pages repo for this organization when it is set up

Comment (dsyer) [inline pom.xml:15]: Actually I think the gh-pages will be `distributedtracing.github.io` so it would be better to start with a maven group `io.github.distributedtracing` if that was the rationale. We should be able to control static content in the domain/subdomain that we claim, so I think ".com"  is wrong.

Comment (rtyler) [inline pom.xml:15]: @dsyer both are equivalent, foo.github.io and foo.github.com are the same.

In the case of a custom CNAME, like we use for @Lookout, lookout.github.io, lookout.github.com and hackers.lookout.com all end up at the same place

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 5 above)
Comment (kristofa) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:3]: Agree. It would be good to add some javadoc here.

**Comment by codefromthecrypt** [comment]
> It looks like most other implementations are familiar with ThreadLocal
> from Java, it's just that we have to sell ContinuationLocal as simply an
> extension to ThreadLocal that also supports thread pools and event loops

Pretty much.

I think that the 'inspired by' story is more a sideline than a headline.
Start with what this code buys you and how to use it (ex htrace homepage
does this well).

Credit inspiration, but at the same time I suspect this will deviate and
that's OK (whereas deviation in clones is usually implicitly not OK).

These suggestions follow normal disclaimer of 'toss if unneeded' ;)

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java [hunk 6]
```diff
@@ -0,0 +1,170 @@
+package com.github.distributedtracing;
```
Comment (louiscryan) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:1]: Is this ultimately what we want? Might as well state your intent and go for

org.continuation

from the start

Comment (b) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:1]: +1 to not having it in the com.github namespace.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+public final class ContinuationLocal<T> {
+    private static InheritableThreadLocal<Object[]> localCtx =
+            new InheritableThreadLocal<Object[]>();
+    private static volatile int size = 0;
+
+    /**
+     * Return a snapshot of the current ContinuationLocal state.
+     */
+    public static Object[] save() {
```
Comment (louiscryan) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:11]: This is likely to violate the mutability expectations of users when state is moved between threads. In particular the fact that some inner code can directly replace the bound value at any time for any ContinuationLocal regardless of visibility rules. I'd strongly consider encapsulating this

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
+package com.github.distributedtracing;
+
+public final class ContinuationLocal<T> {
+    private static InheritableThreadLocal<Object[]> localCtx =
+            new InheritableThreadLocal<Object[]>();
+    private static volatile int size = 0;
```
Comment (louiscryan) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:6]: Better to use AtomicInteger so you don't have to hold the class monitor which risks contention.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+            }
+        };
+    }
+
+    private final int me = add();
+
+    /**
+     * Update the ContinuationLocal with the given value.
+     */
+    public void set(T value) {
```
Comment (louiscryan) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:114]: Is continuous mutability the goal or an anti-pattern here? We explicitly chose not to allow it in

https://github.com/grpc/grpc-java/pull/562

because we believe the aggregate set of 'context' passed to a unit of work should be stable and that it was an exceptional event to need to mutate it. If folks need mutability for some reason they can have the stored state be mutable. 

If your goal is to have the 'feel' of ThreadLocal then it will have to accept the risks that come with that choice.

File: src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
+package com.github.distributedtracing;
```
Comment (louiscryan) [inline src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java:1]: There are two flavors of propagating executors. One which propagates the 'current' state and one which propagates a specific 'saved' state. Should be able to support both flavors here.

File: src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    private Runnable wrapRunnable(final Runnable command) {
+        final Object[] saved = ContinuationLocal.save();
+        return new Runnable() {
+            public void run() {
+                Object[] ctx = ContinuationLocal.save();
+                ContinuationLocal.restore(saved);
+                try {
+                    command.run();
+                } finally {
+                    ContinuationLocal.restore(ctx);
```
Comment (louiscryan) [inline src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java:93]: Another design point which should be explicitly stated here. Do we care if the command is dirty in the sense that it doesn't ensure that the state that was present when it started is the same when it completes.

I agree that you want to forcibly restore the original state in this case but it might be useful to log if saved != current when the command completes.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+                    result = fn.apply();
+                } finally {
+                    ContinuationLocal.restore(save);
+                }
+                return result;
+            }
+        };
+    }
+
+    private final int me = add();
```
Comment (louiscryan) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:109]: While integer keys are efficient the data density tradeoff might be undesirable if folks think that they can create and discard new continuation local instances all the time.

If the performance characteristics of indexed lookup are critical then perhaps place an upper bound on the array size and fallback to a spare map implementation once that size is exceeded.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java [hunk 13]
```diff
@@ -0,0 +1,170 @@
+package com.github.distributedtracing;
+
+public final class ContinuationLocal<T> {
+    private static InheritableThreadLocal<Object[]> localCtx =
```
Comment (louiscryan) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:4]: Is passive inheritance desired? Is it better to be explicit about passing context so that the mechanism by which context carries over thread boundaries is not dependent on how the thread was created. Or to put it another way, everyone should use executors or wrap runnables

Comment (eirslett) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:4]: Are you referring to ThreadLocal vs InheritableThreadLocal? (The only difference being that InheritableThreadLocal passes state from one thread to another thread if the second thread was created from the first thread)

It's a design decision whether or not passing state between threads should be explicit or implicit. The ThreadLocal in Java does it implicitly, so I guess that's the "standard" Java way. Another point of view is that we want to be able to trace requests without having to make lots of changes to userland code. Even more, to instruct developers to manually wrap all their threads and thread pools. In a multi-million-line code base, it's a pain to search-replace every place you need to wrap runnables and so on. (Unfortunately, that's what we have to do anyways, unless the Java standard library adopts something like ContinuationLocal and wraps ExecutorServices natively).

The underlying mechanism for passing state is the save() and restore() stuff, and the ExecutorService wrapper included in this library is simply a default implementation for thread pools. Likewise, InheritableThreadLocal could be thought of as a default implementation for passing state into new threads, I suppose?
Rxjava would have to add save()/restore() hooks to their library in order to achive the same thing.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 12 above)
Comment (eirslett) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:109]: It's right that developers shouldn't make new ContinuationLocals everywhere, because they have an overhead, and aren't really garbage collected either. But I'm not sure the implementation details should document how ContinuationLocal is supposed to be used, maybe that should be a warning in the JavaDoc instead?

File: src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java (hunk 11 above)
Comment (eirslett) [inline src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java:93]: That's a good point! I'm not sure what the correct answer is, but the lifecycle of the command is (hopefully) that it's created once for every submitted task and never reused. Though we cannot guarantee that somebody will reuse a Runnable multiple times.
The problem with logging it, is that we'll have to introduce a dependency on a logging framework... we could maybe use asserts but that will probably break stuff for people who have enabled asserts in their production environment?

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 9 above)
Comment (eirslett) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:114]: That's a very good point! Mutability can be achieved by using a mutable inner value.
I'm not sure what's the right answer here... as you say it's a balance between the ThreadLocal "feel" vs. benefits of immutability. Does anybody else have opinions?

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 8 above)
Comment (eirslett) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:6]: That's a good point! Will change that.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 7 above)
Comment (eirslett) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:11]: Again, also a trade-off here; Twitter solved this by using a type alias:
https://github.com/twitter/util/commit/288c79067c611be5774644cefd17d2c7ddcf28cf
I'm not sure if their construct is 100 % safe either. But the idea was to reduce allocations by not using a wrapper.
I do agree with you that the safety an encapsulating class will provide justifies the allocation overhead. Any other people with opinions?

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 6 above)
Comment (eirslett) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:1]: I do +1 it myself.
org.continuation is a good namespace, since it's not directly tied to distributed tracing. One could use a ContinuationLocal for other things as well, like passing around business logic state, even though I would certainly not recommend that.

File: src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java (hunk 10 above)
Comment (eirslett) [inline src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java:1]: It's the restore(state) method that's used to propagate state, it can accept any state really, as long as it has a reference to it.

File: src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java (hunk 11 above)
Comment (codefromthecrypt) [inline src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java:93]: Wrt logging, just use j.u.Logger? we did that in Brave; it is always there
and much easier on interop that choosing something else, especially when
only used for exception cases.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 7 above)
Comment (adriancole) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java]: Personally, I like the Context impl from gRPC. I think it is a more
idiomatic base to work with, and has had careful review already. We would
get better results starting with that and modifying it vs the current track.

Comment (adriancole) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java]: Just a thought: I think it would be easier to locally store a copy of the
work in progress impl in the intended codebases. Whether this is based on
twitter Local or gRPC Context, it is different than most impls.

What if after a round of feedback, we actually integrate a local beta copy
of the desired source into zipkin or brave etc? After it proves to work,
then formalize and raise a PR to replace it with a dep (along with the
merits of doing so).

The 'if you build it, they will come' approach might work, but proving this
out in practice first wouldn't hurt, and probably reduce thrash. As an
increment step, impls at least have a similar approach to start with and
that can help identify gaps that the common one might need, but are
difficult to guess.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 13 above)
Comment (louiscryan) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:4]: I'm just questioning whether implicit passing via InheritableThreadLocal is a fools errand for the reasons you mention. I consider wrapping Executors to be explicit passing btw and something most modern application do as a matter of routine.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 12 above)
Comment (louiscryan) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:109]: ContinuationLocal instances appear to be garbage collectable (am I missing something?) I agree you want to avoid too much coupling between implementation details and interface but there is a clear risk here. Guidance in doc is probably sufficient for most cases but defense in depth is also worth considering.

File: src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java (hunk 11 above)
Comment (louiscryan) [inline src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java:93]: +1

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 9 above)
Comment (mariusae) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:114]: `com.twitter.util.Local` are not mutable. That is, they are _bindings_ and not _variables_. As @louiscryan points out, you can however bind a ref cell. I think this is the most sensible design.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 13 above)
Comment (ejona86) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:4]: It does seem `InheritableThreadLocal` leads to unpredictability. For instance, `ThreadPoolExecutor.execute()` does create threads, which would lead to the threads in the pool each having a "random" value in the `ContinuationLocal`. It would seem better to obviously break so that the user discovers the missing `wrap()`.

There are also issues with object retention, in that the ThreadPoolExecutor, even when properly wrapped, would be retaining references to the `ContinuationLocal` information from "arbitrary" threads, some of which may be long-dead.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 7 above)
Comment (ejona86) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:11]: It seems dropping some type safety and returning `Object` would be superior over `Object[]`. That 1) allows the concrete implementation to morph and 2) discourages modifying the object. It is true that users could cast to the `Object[]`, but that seems less of a concern. `restore()` would require "saved was returned from save()" and do a cast internally, throwing an `IllegalArgumentException` if it isn't the expected type.

That isn't to say that I think returning an `Object` is better than a special type needing an extra allocation, just that it seems less problematic than `Object[]`.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 12 above)
Comment (eirslett) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:109]: The objects that are referred to by a ContinuationLocal is stored inside the context (the current/temporary implementation of context is Object[]). So the ContinuationLocal itself is garbage collected, but not the value it's bound to.
Anyways, I suppose the "correct" use of a ContinuationLocal is to have it in a global, static variable? So it will have the same lifecycle as the application itself. Are there any cases where one would need additional scoping?

File: src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java (hunk 11 above)
Comment (eirslett) [inline src/main/java/com/github/distributedtracing/ContinuationLocalExecutorService.java:93]: That's a good point! j.u.Logger is an alternative here.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 9 above)
Comment (eirslett) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:114]: @mariusae wouldn't this set() method on the Local mean it's mutable in the same sense that a ThreadLocal is mutable?
https://github.com/twitter/util/blob/master/util-core/src/main/scala/com/twitter/util/Local.scala#L117

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 7 above)
Comment (eirslett) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:11]: That's a good solution!

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 13 above)
Comment (eirslett) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:4]: @mariusae was it a conscious decision to use ThreadLocal insted of InheritableThreadLocal, in the first place?

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 9 above)
Comment (mariusae) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:114]: It's really hard to compare, because ThreadLocals don't propagate across their delimitation's boundaries (starting a new thread) whereas Locals do.

Thus, if I change a Local with `set()`, it doesn't affect any context where those Locals have been suspended.

Even so, `set` is a low-level API, and user code should instead use `let` which emphasizes the binding nature of the context. (This should be made more clear in its documentation.) https://github.com/twitter/util/blob/master/util-core/src/main/scala/com/twitter/util/Local.scala#L128

Comment (mariusae) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:114]: Also, the implication of this, is that if you _do_ want to share values that mutate across contexts, you do have to use inner refs.

Finagle's context API provides a clearer API for all of this. (It does more--but it might inform how I'd write `com.twitter.util.Local` were I to do it again.) https://github.com/twitter/finagle/blob/develop/finagle-core/src/main/scala/com/twitter/finagle/context/Context.scala

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 13 above)
Comment (mariusae) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:4]: @eirslett yes. we wanted all context saving and restoring to be explicit.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+    /**
+     * Restore the ContinuationLocal state to a given set of values.
+     */
+    public static void restore(Object[] saved) {
+        localCtx.set(saved);
+    }
+
+    private static synchronized int add() {
+        size += 1;
+        return size - 1;
```
Comment (mariusae) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:24]: return size++?

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+     */
+    private static void clearAll() {
+        localCtx.set(null);
+    }
+
+    /**
+     * Execute a block with the given ContinuationLocals,
+     * restoring current values upon completion.
+     */
+    public static <U> U let(Object[] ctx, Producer<U> f) {
```
Comment (mariusae) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:67]: this is not the usual meaning of `let`.

File: src/main/java/com/github/distributedtracing/ContinuationLocal.java (hunk 12 above)
Comment (mariusae) [inline src/main/java/com/github/distributedtracing/ContinuationLocal.java:109]: Just as a data point: we're operating really rather huge codebases with `com.twitter.util.Local`, and we've never had any issues with this approach.

**Comment by schlosna** [comment]
I'm curious what the next steps are here to get this merging, or if this is effectively at dead PR. I'm mostly interested in this in the context of https://github.com/openzipkin/brave/issues/166 so if there is anything I can do to help push this, please let me know. Thanks!

**Comment by codefromthecrypt** [comment]
Jakub was looking at this, but offline. If you are in a position to tackle
this in Brave, I'd go for it. Many people would like this solved including
me.

**Comment by schlosna** [comment]
Thanks Adrian. I actually work with Jakub, and have some time this week as part of a hack week project instrumenting a variety of systems with Zipkin & Brave, so I'll create a Brave PR to iterate.

**Comment by eirslett** [comment]
You're free to pick it up from here!

**Comment by codefromthecrypt** [comment]
Keep us posted on gitter! Ps once things are sorted, local to brave, we can
also revisit this PR with experiences from it.

**Comment by codefromthecrypt** [comment]
some interesting development out of this is a recent addition to log4j 2.7 which incidentally came out of a feature request to pull in state from finagle locals. https://issues.apache.org/jira/browse/LOG4J2-1010

I'd say [ContextDataInjector](https://logging.apache.org/log4j/2.x/log4j-core/apidocs/org/apache/logging/log4j/core/ContextDataInjector.html) is an example of something we could attempt to implement in a usage test case, to see if we can share a context across diversely written apps.

**Comment by codefromthecrypt** [comment]
ps thanks for this.. I've grabbed the parts needed for brave. we can revisit this in the new design as context is pluggable now

https://github.com/openzipkin/brave/pull/369
